gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -210,22 +213,65 @@ func (vm *VMKeeper) newGnoTransactionStore(ctx sdk.Context) gno.TransactionStore
 }
 
 func (vm *VMKeeper) MakeGnoTransactionStore(ctx sdk.Context) sdk.Context {
-	return ctx.WithValue(gnoStoreContextKey, vm.newGnoTransactionStore(ctx))
+	return ctx.WithValue(keeperKeyGnoStore, vm.newGnoTransactionStore(ctx))
 }
 
 func (vm *VMKeeper) CommitGnoTransactionStore(ctx sdk.Context) {
 	vm.getGnoTransactionStore(ctx).Write()
 }
 
 func (vm *VMKeeper) getGnoTransactionStore(ctx sdk.Context) gno.TransactionStore {
-	txStore := ctx.Value(gnoStoreContextKey).(gno.TransactionStore)
+	txStore := ctx.Value(keeperKeyGnoStore).(gno.TransactionStore)
 	txStore.ClearObjectCache()
 	return txStore
 }
 
 // Namespace can be either a user or crypto address.
 var reNamespace = regexp.MustCompile(`^gno.land/(?:r|p)/([\.~_a-zA-Z0-9]+)`)
 
+// ExecContextCustom is a subset of [stdlibs.ExecContext] which can be used to
+// inject custom fields at genesis.
+type ExecContextCustom struct {
+	// Height is the BlockHeight: it is not recommended to set the same
+	// BlockHeight as the one used by the previous iteration of the chain, as
+	// from the realm's point of view it will eventually be as if a height
+	// occurred twice at different times: instead, use negative heights.
+	// (Last block should be -1, as 0 is the genesis block).
+	Height    int64 `json:"height"`
+	Timestamp int64 `json:"timestamp"`
```

Comment:
Is it required to have both height and timestamp? Could we just have height and derive timestamps for each height? There could be some wonky behavior here if the tx at height -4 has a timestamp after the tx at height -1, or swap height for timestamp if we choose it replay transactions by timestamp rather than height.

Reply:
Ideally not.

Timestamp != `genesisTime + blockHeight * blockTime`, because there may be multiple consensus rounds; so it has to be incremented by `failedRounds * blockTime` at least.

So I'd keep them separate.

Reply:
I understand that is not how timestamp is computed, but what do failed rounds have to do with executing transactions with negative heights? Won't they be executed before genesis? Are there consensus rounds for this? When I asked about deriving timestamps, I was referring to deriving timestamps for transactions with negative heights.  I'm just trying to understand the negative heights and how or if there is any relationship between negative height and timestamp and if a transaction with an earlier timestamp can be executed before a transaction with a later timestamp.

Reply:
>When I asked about deriving timestamps, I was referring to deriving timestamps for transactions with negative heights. I'm just trying to understand the negative heights and how or if there is any relationship between negative height and timestamp and if a transaction with an earlier timestamp can be executed before a transaction with a later timestamp.

So, we likely won't be going down the road of negative block heights (the current state of this PR), but I'll still try to answer your question.

This PR per se does not define when a transaction with negative height should happen. It could be at any point in time; the only consequence is, of course, that some realms will expect ascending order for both timestamps and block heights, and have code work under this assumption.

For the use case I envisioned of this feature, though, I wanted to have the result of `time.Now()` essentially be always the same on the portal loop. So that your transaction is first run on the portal loop at `height=10 time=2024-09-12T18:30:00Z`; then when the portal loop restarts and all the transactions are dumped into the genesis file, the transaction is run again with `height=-X time=2024-09-12T18:30:00Z`; although it's not the "real" block time and height, but rather an emulated one, injected through the `TxContexts` field in genesis.json